File: javascript/grains/grains.js
import BigInteger from './lib/big-integer';

export const numOfGrains = (x) => BigInteger(2).pow(x - 1)

export const square = (x) => {
  if (x < 1 || x > 64) throw new Error('square must be between 1 and 64')
  return numOfGrains(x).toString()
}

export const total = () => numOfGrains(65).minus(1).toString()

File: haskell/grains/src/Grains.hs
module Grains (square, total) where

numOfGrains :: Integer -> Integer
numOfGrains n = 2^(n - 1)

square :: Integer -> Maybe Integer
square n
  | n > 64 || n < 1 = Nothing
  | otherwise = Just (numOfGrains n)

total :: Integer
total = (numOfGrains 65) - 1
